security/pkg/nodeagent/sds/sdsservice.go

```diff
@@ -18,576 +18,167 @@ package sds
-	}
-
-	sdsClientsMutex.Lock()
-	defer sdsClientsMutex.Unlock()
-
-	// Only add connection key to staledClientKeys if it's not there already.
-	// The recycleConnection function may be triggered more than once for each connection key.
-	// https://github.com/istio/istio/issues/15306#issuecomment-509783105
-	if _, found := staledClientKeys[key]; found {
-		return
-	}
-
-	staledClientKeys[key] = true
-
-	totalStaleConnCounts.Increment()
+// register adds the SDS handle to the grpc server
+func (s *sdsservice) register(rpcs *grpc.Server) {
+	sds.RegisterSecretDiscoveryServiceServer(rpcs, s)
 }
 
-func getResourceName(discReq *discovery.DiscoveryRequest) (string /*resourceName*/, error) {
-	if len(discReq.ResourceNames) == 0 {
-		return "", nil
-	}
-
-	if len(discReq.ResourceNames) == 1 {
-		return discReq.ResourceNames[0], nil
-	}
-
-	return "", fmt.Errorf("discovery request %+v has more than one resourceNames %+v", discReq, discReq.ResourceNames)
+// StreamSecrets serves SDS discovery requests and SDS push requests
+func (s *sdsservice) StreamSecrets(stream sds.SecretDiscoveryService_StreamSecretsServer) error {
+	return s.XdsServer.Stream(stream)
 }
 
-func getCredentialToken(ctx context.Context) (string, error) {
-	metadata, ok := metadata.FromIncomingContext(ctx)
-	if !ok {
-		return "", fmt.Errorf("unable to get metadata from incoming context")
-	}
-
-	// Get credential token from request k8sSAJwtTokenHeader(`istio_sds_credentail_header`) if it exists;
-	// otherwise fallback to credentialTokenHeader('authorization').
-	if h, ok := metadata[k8sSAJwtTokenHeaderKey]; ok {
-		if len(h) != 1 {
-			return "", fmt.Errorf("credential token from %q must have 1 value in gRPC metadata but got %d", k8sSAJwtTokenHeaderKey, len(h))
-		}
-		if len(h[0]) == 0 {
-			return "", fmt.Errorf("received empty credential for existing header: %s", k8sSAJwtTokenHeaderKey)
-		}
-		return h[0], nil
-	}
-
-	if h, ok := metadata[credentialTokenHeaderKey]; ok {
-		if len(h) != 1 {
-			return "", fmt.Errorf("credential token from %q must have 1 value in gRPC metadata but got %d", credentialTokenHeaderKey, len(h))
-		}
-		if len(h[0]) == 0 {
-			return "", fmt.Errorf("received empty credential for existing header: %s", credentialTokenHeaderKey)
-		}
-		return h[0], nil
-	}
-
-	return "", fmt.Errorf("no credential token is found")
+func (s *sdsservice) DeltaSecrets(stream sds.SecretDiscoveryService_DeltaSecretsServer) error {
+	return status.Error(codes.Unimplemented, "DeltaSecrets not implemented")
 }
 
-func addConn(k cache.ConnKey, conn *sdsConnection) {
-	sdsClientsMutex.Lock()
-	defer sdsClientsMutex.Unlock()
-	conIDresourceNamePrefix := sdsLogPrefix(k.ResourceName)
-	sdsServiceLog.Debugf("%s add a new connection", conIDresourceNamePrefix)
-	sdsClients[k] = conn
+func (s *sdsservice) FetchSecrets(ctx context.Context, discReq *discovery.DiscoveryRequest) (*discovery.DiscoveryResponse, error) {
+	return nil, status.Error(codes.Unimplemented, "FetchSecrets not implemented")
```

Comment:
Can we keep the implementation of FetchSecrets()?

Reply:
Why? It has not been used by envoy since xds v2 which has been a very long time